funcx_endpoint/funcx_endpoint/endpoint/endpoint.py

```diff
@@ -308,46 +309,50 @@ def start_endpoint(
         os.path.join(endpoint_dir, FUNCX_CONFIG_FILE_NAME)).load_module()
 
     with context:
-        while True:
-            # Register the endpoint
-            logger.info("Registering endpoint")
-            if State.FUNCX_CONFIG.get('broker_test', False) is True:
-                logger.warning("**************** BROKER State.DEBUG MODE *******************")
-                reg_info = register_with_hub(endpoint_uuid,
-                                             endpoint_dir,
-                                             State.FUNCX_CONFIG['broker_address'],
-                                             client_public_key=client_public_key,
-                                             redis_host=State.FUNCX_CONFIG['redis_host'])
-            else:
-                reg_info = register_endpoint(funcx_client, name, endpoint_uuid, endpoint_dir)
-
-            logger.info("Endpoint registered with UUID: {}".format(reg_info['endpoint_id']))
-
-            # Configure the parameters for the interchange
-            optionals = {}
-            optionals['client_address'] = reg_info['public_ip']
-            optionals['client_ports'] = reg_info['tasks_port'], reg_info['results_port'], reg_info['commands_port'],
-            if 'endpoint_address' in State.FUNCX_CONFIG:
-                optionals['interchange_address'] = State.FUNCX_CONFIG['endpoint_address']
-
-            optionals['logdir'] = endpoint_dir
-
-            if State.DEBUG:
-                optionals['logging_level'] = logging.DEBUG
+        reg_info = handle_start_registration(funcx_client, name, endpoint_uuid, endpoint_dir)
```

Comment:
There are two situations that need to be considered here: 1) User just requested an endpoint start, and registration failed 2) EndpointInterchange failed due to a disconnect, and we are back at endpoint register due to the retry decorator.

In case 1, we have an exception that we could present to the user immediately, and it's useful to do that since this might be from a configuration error than the user needs to fix.
In case 2, we know that the configured values worked at some point, so retrying periodically is not a bad idea.

For case 1 to work, handle_start_registration needs to raise an exception, and we need to report it right away if this is the first time handle_start_registration has been called.
For case 2, we need some mechanism to figure out if the EndpointInterchange worked at all, right now we should assume that if there were no exception raise during EI instantiation, it connected. This could be improved later.

Reply:
To handle case 1, as we discussed over yesterdays's group meeting, introduced a new variable `previously_registered`, which is the existence of `endpoint.json` file, and defines if the endpoint has been registered successfully before. If `previously_registered` is `False`, retry will not happen. After printing out a warning message, the program would terminate.